// genezio/build/variables.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LOCAL_TEST_INTERFACE_URL = exports.ENABLE_DEBUG_LOGS_BY_DEFAULT = exports.PORT_LOCAL_ENVIRONMENT = exports.GENERATE_SDK_API_URL = exports.BACKEND_ENDPOINT = exports.FRONTEND_DOMAIN = exports.REACT_APP_BASE_URL = void 0;
const environment = "production";
let REACT_APP_BASE_URL;
exports.REACT_APP_BASE_URL = REACT_APP_BASE_URL;
let BACKEND_ENDPOINT;
exports.BACKEND_ENDPOINT = BACKEND_ENDPOINT;
let FRONTEND_DOMAIN;
exports.FRONTEND_DOMAIN = FRONTEND_DOMAIN;
let GENERATE_SDK_API_URL;
exports.GENERATE_SDK_API_URL = GENERATE_SDK_API_URL;
let PORT_LOCAL_ENVIRONMENT;
exports.PORT_LOCAL_ENVIRONMENT = PORT_LOCAL_ENVIRONMENT;
let LOCAL_TEST_INTERFACE_URL;
exports.LOCAL_TEST_INTERFACE_URL = LOCAL_TEST_INTERFACE_URL;
let ENABLE_DEBUG_LOGS_BY_DEFAULT;
exports.ENABLE_DEBUG_LOGS_BY_DEFAULT = ENABLE_DEBUG_LOGS_BY_DEFAULT;
if (environment === "dev") {
    exports.REACT_APP_BASE_URL = REACT_APP_BASE_URL = "https://dev.app.genez.io";
    exports.FRONTEND_DOMAIN = FRONTEND_DOMAIN = "dev.app.genez.io";
    exports.BACKEND_ENDPOINT = BACKEND_ENDPOINT = "https://dev.api.genez.io";
    exports.GENERATE_SDK_API_URL = GENERATE_SDK_API_URL = "https://dev-sdk-api.genez.io";
    exports.LOCAL_TEST_INTERFACE_URL = LOCAL_TEST_INTERFACE_URL = "https://dev.app.genez.io/test-interface/local";
    exports.PORT_LOCAL_ENVIRONMENT = PORT_LOCAL_ENVIRONMENT = 8083;
    exports.ENABLE_DEBUG_LOGS_BY_DEFAULT = ENABLE_DEBUG_LOGS_BY_DEFAULT = true;
}
else {
    exports.REACT_APP_BASE_URL = REACT_APP_BASE_URL = "https://app.genez.io";
    exports.FRONTEND_DOMAIN = FRONTEND_DOMAIN = "app.genez.io";
    exports.BACKEND_ENDPOINT = BACKEND_ENDPOINT = "https://api.genez.io";
    exports.GENERATE_SDK_API_URL = GENERATE_SDK_API_URL = "https://sdk-api.genez.io";
    exports.LOCAL_TEST_INTERFACE_URL = LOCAL_TEST_INTERFACE_URL = "https://app.genez.io/test-interface/local";
    exports.PORT_LOCAL_ENVIRONMENT = PORT_LOCAL_ENVIRONMENT = 8083;
    exports.ENABLE_DEBUG_LOGS_BY_DEFAULT = ENABLE_DEBUG_LOGS_BY_DEFAULT = false;
}
