// genezio/build/requests/axios.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const axios_1 = __importDefault(require("axios"));
const logging_1 = require("../utils/logging");
var GenezioErrorCode;
(function (GenezioErrorCode) {
    GenezioErrorCode[GenezioErrorCode["UnknownError"] = 0] = "UnknownError";
    GenezioErrorCode[GenezioErrorCode["Unauthorized"] = 1] = "Unauthorized";
    GenezioErrorCode[GenezioErrorCode["NotFoundError"] = 2] = "NotFoundError";
    GenezioErrorCode[GenezioErrorCode["InternalServerError"] = 3] = "InternalServerError";
    GenezioErrorCode[GenezioErrorCode["MethodNotAllowedError"] = 4] = "MethodNotAllowedError";
    GenezioErrorCode[GenezioErrorCode["MissingRequiredParametersError"] = 5] = "MissingRequiredParametersError";
    GenezioErrorCode[GenezioErrorCode["BadRequest"] = 6] = "BadRequest";
    GenezioErrorCode[GenezioErrorCode["StatusConflict"] = 7] = "StatusConflict";
    GenezioErrorCode[GenezioErrorCode["UpdateRequired"] = 8] = "UpdateRequired";
})(GenezioErrorCode || (GenezioErrorCode = {}));
axios_1.default.interceptors.response.use(function (response) {
    // Do something with response data
    return response;
}, function (error) {
    var _a, _b, _c, _d, _e;
    if (((_c = (_b = (_a = error.response) === null || _a === void 0 ? void 0 : _a.data) === null || _b === void 0 ? void 0 : _b.error) === null || _c === void 0 ? void 0 : _c.code) === GenezioErrorCode.UpdateRequired) {
        throw new Error("Please update your genezio CLI. Run 'npm update -g genezio'.");
    }
    else if ((_e = (_d = error.response) === null || _d === void 0 ? void 0 : _d.data) === null || _e === void 0 ? void 0 : _e.error) {
        logging_1.debugLogger.debug(JSON.stringify(error.response.data.error));
    }
    else {
        logging_1.debugLogger.debug(error);
    }
    // Do something with response error
    return Promise.reject(error);
});
exports.default = axios_1.default;
