// genezio/build/requests/deployCode.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.deployRequest = void 0;
const axios_1 = __importDefault(require("./axios"));
const variables_1 = require("../variables");
const accounts_1 = require("../utils/accounts");
const logging_1 = require("../utils/logging");
const logging_2 = require("../utils/logging");
const node_abort_controller_1 = require("node-abort-controller");
const strings_1 = require("../utils/strings");
// eslint-disable-next-line @typescript-eslint/no-var-requires
const pjson = require("../../package.json");
function deployRequest(projectConfiguration) {
    var _a, _b;
    return __awaiter(this, void 0, void 0, function* () {
        // auth token
        (0, logging_2.printAdaptiveLog)("Checking your credentials", "start");
        const authToken = yield (0, accounts_1.getAuthToken)();
        if (!authToken) {
            (0, logging_2.printAdaptiveLog)("Checking your credentials", "error");
            throw new Error("You are not logged in. Run 'genezio login' before you deploy your function.");
        }
        (0, logging_2.printAdaptiveLog)("Checking your credentials", "end");
        const json = JSON.stringify({
            classes: projectConfiguration.classes,
            projectName: projectConfiguration.name,
            region: projectConfiguration.region,
            cloudProvider: projectConfiguration.cloudProvider,
        });
        const controller = new node_abort_controller_1.AbortController();
        const messagePromise = (0, logging_2.printUninformativeLog)(controller);
        const response = yield (0, axios_1.default)({
            method: "PUT",
            url: `${variables_1.BACKEND_ENDPOINT}/core/deployment`,
            data: json,
            headers: {
                Authorization: `Bearer ${authToken}`,
                "Accept-Version": `genezio-cli/${pjson.version}`
            },
            maxContentLength: Infinity,
            maxBodyLength: Infinity
        }).catch((error) => __awaiter(this, void 0, void 0, function* () {
            controller.abort();
            (0, logging_2.printAdaptiveLog)(yield messagePromise, "error");
            logging_1.debugLogger.debug("Error received", error);
            throw error;
        }));
        controller.abort();
        (0, logging_2.printAdaptiveLog)(yield messagePromise, "end");
        logging_1.debugLogger.debug("Response received", response.data);
        if (response.data.status === "error") {
            if (response.data.message === "Unauthorized") {
                throw new Error(strings_1.GENEZIO_NOT_AUTH_ERROR_MSG);
            }
            throw new Error(response.data.message);
        }
        if ((_b = (_a = response.data) === null || _a === void 0 ? void 0 : _a.error) === null || _b === void 0 ? void 0 : _b.message) {
            throw new Error(response.data.error.message);
        }
        return response.data;
    });
}
exports.deployRequest = deployRequest;
